### PR TITLE
Initial version

### PR DESCRIPTION
**Pull request by bhargavprajapatiFynd**

### CODE REVIEW

File: tests/src/utils.spec.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import querystring from "querystring";
+
+describe("Test hmac", () => {
+    it("Should encrypt string using key", async () => {
+        const key = "keyToEncrypt";
+        const stringToEncrypt = "This is secret message to encrypt";
+
+        const expectedEncryption = "baa3f0a60d6cdb60e23aa7c5369e3a27e11632248d56e752cc96cafa7ade5aec";
+
+        const encryptedMsg = hmac(key, stringToEncrypt, "");
```
[the remark this answers is not in the document]
Comment (bhargavprajapatiFynd) [reply tests/src/utils.spec.ts]: `encodeRfc3986` is not a library function. It is defined in `src/utils.ts`

File: package.json [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    "build": "npm run clear && concurrently 'npm:build:cjs' 'npm:build:mjs' 'npm:build:types' 'npm:build:web'",
+    "build:cjs": "tsc -p tsconfig-cjs.json",
+    "build:mjs": "tsc -p tsconfig-mjs.json",
+    "build:web": "webpack",
+    "build:types": "tsc -p tsconfig-types.json",
+    "prepublishOnly": "npm run build",
+    "clear": "rimraf ./dist",
+    "test": "jest --coverage"
+  },
+  "repository": {
```
Comment (vivek-gofynd) [inline package.json:19]: Add keywords in package.json @bhargavprajapatiFynd

File: README.md [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+### For ES Modules
+
+```javascript
+import { sign } from "@gofynd/fp-signature";
+```
+
+### For Browser
+
+```html
+<script src="https://cdn.jsdelivr.net/npm/@gofynd/fp-signature"></script>
```
Comment (vivek-gofynd) [inline README.md:34]: Are we publishing this on CDN ? @bhargavprajapatiFynd

File: README.md [hunk 4]
```diff
@@ -0,0 +1,97 @@
...
+    // FP-Signature library will be attached to the global window object
+    FPSignature.sign();
+</script>
+```
+
+## Functionality
+
+### `sign` function
+
+The `sign` function is used to add a signature to a request. It takes two parameters: `request` and `options`.
```
Comment (vivek-gofynd) [inline README.md:45]: Change wordings here it will return generated signature

File: README.md (hunk 3 above)
Comment (bhargavprajapatiFynd) [reply README.md:34]: yes
cdn url provided in README is working

File: src/RequestSigner.ts [hunk 5]
```diff
@@ -0,0 +1,296 @@
...
+
+export type RequestParam = {
+  headers?: any;
+  method: string;
+  host?: string;
+  hostname?: string;
+  port?: number;
+  path?: string;
+  body?: any;
+  service?: any;
```
[the remark this answers is not in the document]
Comment (bhargavprajapatiFynd) [reply src/RequestSigner.ts]: done

File: README.md (hunk 4 above)
Comment (bhargavprajapatiFynd) [reply README.md]: changed line

File: package.json [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+  },
+  "homepage": "https://github.com/gofynd/fp-signature.git",
+  "author": {
+    "name": "Fynd Developer",
+    "email" : "dev@fynd.com"
+  },
+  "license": "ISC",
+  "dependencies": {
+    "crypto-js": "^4.1.1",
+    "querystring-es3": "^0.2.1"
```
Comment (vivek-gofynd) [inline package.json:31]: check if this can be moved into dev dependencies @bhargavprajapatiFynd

File: package.json (hunk 2 above)
Comment (bhargavprajapatiFynd) [reply package.json:19]: Added keywords

File: package.json (hunk 6 above)
Comment (bhargavprajapatiFynd) [reply package.json]: Moved querystring-es3 to devdependency

File: README.md [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+  path?: string;
+  headers?: any;
+  body?: any;
+  doNotEncodePath?: boolean;
+  doNotModifyHeaders?: boolean; 
+};
+
+type SigningOptions = {
+    secret?: string;
+    forQuery?: boolean;
```
Comment (vivek-gofynd) [inline README.md:62]: change this option to signQuery @bhargavprajapatiFynd

File: README.md [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    method: "GET",
+    host: "api.fynd.com",
+    path: "/service/application/configuration/v1.0/application",
+    headers: {
+      Authorization: "Bearer NjQ2NGIxMTY5YThjNmI3ZDUwMDVkMTJlOnVfeXhsWXBaQg==",
+      "x-currency-code": "INR"
+    },
+};
+
+const signature = sign(requestToSign)
```
Comment (vivek-gofynd) [inline README.md:142]: what this function will return and how to use it, add both exmaples one for query and one for header. @bhargavprajapatiFynd

File: README.md (hunk 7 above)
Comment (bhargavprajapatiFynd) [reply README.md]: Changed the name

File: README.md (hunk 8 above)
Comment (bhargavprajapatiFynd) [reply README.md]: Solved

**Review by vivek-gofynd — APPROVED** [review]

**Pull request merged**